.github/workflows/benchmarks.yml

```diff
@@ -0,0 +1,164 @@
+    outputs:
+      benchmark-matrix: ${{ steps.export.outputs.benchmark-matrix }}
+      benchmark-config: ${{ steps.export.outputs.benchmark-config }}
+      benchmark-config-gcs-artifact: ${{ steps.upload.outputs.benchmark-config-gcs-artifact }}
+    steps:
+      - name: "Checking out repository"
+        uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # v2
+      - name: "Checking out runtime submodules"
+        run: ./build_tools/scripts/git/update_runtime_submodules.sh
+      - name: "Exporting benchmark run config"
+        id: export
+        env:
+          BENCHMARK_CONFIG: "benchmark-config.json"
+          BENCHMARK_PRESETS: ${{ inputs.benchmark-presets }}
+        run: |
+          ./build_tools/benchmarks/export_benchmark_config.py \
+            --benchmark_presets="${BENCHMARK_PRESETS}" \
+            --output="${BENCHMARK_CONFIG}"
+          echo "benchmark-config=${BENCHMARK_CONFIG}" >> "${GITHUB_OUTPUT}"
+          echo benchmark-matrix=$(jq \
+              'to_entries | map({"device_name": .key, "host_environment": .value.host_environment})' \
+              "${BENCHMARK_CONFIG}") \
+            >> "${GITHUB_OUTPUT}"
+      - name: "Uploading benchmark config"
+        id: upload
+        env:
+          BENCHMARK_CONFIG: ${{ steps.export.outputs.benchmark-config }}
+          BENCHMARK_CONFIG_GCS_ARTIFACT: ${{ env.GCS_DIR }}/${{ steps.export.outputs.benchmark-config }}
+        run: |
+          gcloud alpha storage cp \
+            "${BENCHMARK_CONFIG}" \
+            "${BENCHMARK_CONFIG_GCS_ARTIFACT}"
+          echo "benchmark-config-gcs-artifact=${BENCHMARK_CONFIG_GCS_ARTIFACT}" >> "${GITHUB_OUTPUT}"
+
+  run_benchmarks:
+    needs: [export_benchmark_config]
+    strategy:
+      matrix:
+        benchmark: ${{ fromJSON(needs.export_benchmark_config.outputs.benchmark-matrix) }}
+    runs-on:
+      - self-hosted  # must come first
+      - runner-group=${{ inputs.runner-group }}
+      - environment=${{ inputs.runner-env }}
+      - machine-type=${{ matrix.benchmark.device_name }}
+    outputs:
+      benchmark-results-gcs-artifact-dir: ${{ steps.upload.outputs.benchmark-results-gcs-artifact-dir }}
+    env:
+      DEVICE_NAME: ${{ matrix.benchmark.device_name }}
+      PLATFORM_ARCH: ${{ matrix.benchmark.host_environment.platform }}-${{ matrix.benchmark.host_environment.architecture }}
+      E2E_TEST_ARTIFACTS_GCS_ARTIFACT_DIR: ${{ inputs.e2e-test-artifacts-gcs-artifact-dir }}
+      E2E_TEST_ARTIFACTS_DIR: ${{ inputs.e2e-test-artifacts-dir }}
+    steps:
+      - name: "Checking out repository"
+        uses: actions/checkout@7884fcad6b5d53d10323aee724dc68d8b9096a2e # v2
+      - name: "Checking out runtime submodules"
+        run: ./build_tools/scripts/git/update_runtime_submodules.sh
+      - name: "Downloading benchmark tools"
+        id: download-tools
+        env:
+          # See `build_benchmark_tools` step in ci.yml for the name format of
+          # benchmark tools artifacts.
```

Comment:
Tried but didn't find a way to export a matrix output from a matrix job (`build_benchmark_tools`), so I constructed the file path manually here instead of passing from the job.